src/clojuratica/base/parse.clj

```diff
@@ -1,58 +1,62 @@
+   [clojuratica.jlink]
+   [clojuratica.lib.options :as options]
+   [clojuratica.base.expr :as expr]
+   [clojure.string :as str])
   (:import [com.wolfram.jlink Expr]))
 
-(in-ns 'clojuratica.base.cep)
-(declare cep)
-(in-ns 'clojuratica.base.parse)
-;; (require '[clojuratica.base.cep :as cep])
-;(refer 'clojuratica.base.cep)
-
-(declare
-  parse
-  simple?
-  atom?
-  simple-array-type
-  simple-vector-type
-  simple-matrix-type
-  bound-map
-  vec-bound-map
-  seq-bound-map
-  seq-fn-bound-map
-  parse-simple-vector
-  parse-simple-matrix
-  parse-simple-atom
-  parse-complex-atom
-  parse-complex-list
-  parse-integer
-  parse-rational
-  parse-symbol
-  parse-hash-map
-  parse-fn
-  parse-generic-expression)
-
-(defn parse [expr]
-  (assert (instance? com.wolfram.jlink.Expr expr))
-  (cond (options/flag? dynamic-vars/*options* :as-function)                   (parse-fn expr)
-        (or (atom? expr) (options/flag? dynamic-vars/*options* :full-form))   (parse-complex-atom expr)
-        (simple-vector-type expr)                        (parse-simple-vector expr)
-        (simple-matrix-type expr)                        (parse-simple-matrix expr)
-        :else                                            (parse-complex-list expr)))
-
-(defn simple? [expr]
-  (or (atom? expr) (simple-array-type expr)))
+(declare parse)
+
+(defn pascal->kebab
+  [s]
+  (some-> s
+          (str/replace #"\w(\p{Upper})" (fn [[[p l] _]] (str  p "-" (str/lower-case l))))
+          str/lower-case))
+
+(defn entity-type->keyword [expr opts]
+  (let [head (pascal->kebab (expr/head-str expr))
+        parts (reduce (fn [acc part]
+                        (into acc (cond
+                                    (string? part) [(pascal->kebab part)]
+                                    (coll? part) (reverse (map pascal->kebab part))
+                                    :else nil)))
+                      [head]
+                      (map #(parse % opts) (.args expr)))]
+    ;; BAIL: if any of the parts are not reckognized,
+    ;; let it go through normal parsing
+    (when (not-any? nil? parts)
+      (keyword (str/join "." (butlast parts))
+               (last parts)))))
+
+(defn custom-parse-dispatch [expr {:keys [parse/custom-parse-symbols]}]
+  (let [head (symbol (expr/head-str expr))]
+    (when-not
+        (and (seq custom-parse-symbols) (not (contains? (set custom-parse-symbols) head)))
+      head)))
+
+(ns-unmap *ns* 'custom-parse)
+(defmulti custom-parse #'custom-parse-dispatch)
+
+;; (defmethod custom-parse 'Hyperlink [expr opts]
+;;   (let [parsed-url (parse (second (.args expr)) opts)]
+;;     (try
+;;       (java.net.URL. parsed-url)
+;;       (catch java.net.MalformedURLException _
+;;         parsed-url))))
+
+(defmethod custom-parse 'EntityProperty [expr opts]
+  (entity-type->keyword expr opts))
+
+(defmethod custom-parse 'Entity [expr opts]
+  (entity-type->keyword expr opts))
+
+(defmethod custom-parse :default [_ _]
+  ::not-implemented)

```

Comment:
A new multimethod `custom-parse` is introduced, to open things up for user extension 

the commented out example is a good one i.e. you might want to parse a WL `Hyperlink` entity into a specific thing e.g. `java.net.url`